dnf/transaction_sr.py

```diff
@@ -0,0 +1,169 @@
+        """
+        if isinstance(errors, (list, tuple)):
+            if len(errors) > 1:
+                msg = _('Errors in "{filename}":').format(filename=filename)
+                for error in errors:
+                    msg += "\n  " + str(error)
+
+                super(TransactionFileError, self).__init__(msg)
+                return
+
+            else:
+                errors = str(errors[0])
+
+        msg = _('Error in "{filename}": {error}').format(filename=filename, error=errors)
+        super(TransactionFileError, self).__init__(msg)
+
+
+class IncompatibleTransactionVersionError(TransactionFileError):
+    def __init__(self, filename, msg):
+        super(IncompatibleTransactionVersionError, self).__init__(filename, msg)
+
+
+def check_version(version, filename):
+    major, minor = version.split('.')
+    if major != VERSION_MAJOR:
+        raise IncompatibleTransactionVersionError(
+            filename,
+            _('Incompatible major version {major}, supported major version is {major_supp}.')
+                .format(major=major, major_supp=VERSION_MAJOR)
+        )
+
+    if minor < VERSION_MINOR:
+        raise IncompatibleTransactionVersionError(
+            filename,
+            _('Incompatible minor version {minor}, '
+                'supported minor version is {minor_supp} or greater.'
+            ).format(minor=minor, minor_supp=VERSION_MINOR)
+        )
+
+
+def store_transaction(transaction, filename="transaction.json"):
+    data = {
+        "version": VERSION,
+    }
+    rpms = {}
+    groups = {}
+    environments = {}
+
+    for tsi in transaction.packages():
+        if tsi.is_package():
+            na = "%s.%s" % (tsi.name, tsi.arch)
+
+            data_p = rpms.setdefault(na, {})
+
+            # for installonly packages the package data needs to be a list of actions
+            if isinstance(data_p, list):
+                data_p.append({})
+                data_p = data_p[-1]
+
+            if tsi._item.isForwardAction():
+                if "to" in data_p or "action" in data_p and data_p["action"] == "Removed":
+                    # installonly pkg, convert to a list of actions
+                    rpms[na] = [data_p, {}]
+                    data_p = rpms[na][-1]
+
+                data_p["to"] = tsi.evr
+                data_p["action"] = tsi.action_name
+            elif tsi._item.isBackwardAction():
+                if "from" in data_p or "action" in data_p and data_p["action"] == "Install":
+                    # installonly pkg, convert to a list of actions
+                    rpms[na] = [data_p, {}]
+                    data_p = rpms[na][-1]
+
+                data_p["from"] = tsi.evr
+                # make sure not to overwrite the forward action if it was written already
+                if "action" not in data_p:
+                    data_p["action"] = tsi.action_name
+            else:
+                data_p["weird_reason"] = True
```

Comment:
I am sorry, but this is not a nice way to record something that something is wrong.

Reply:
I forgot to address this, sorry, it's obviously a leftover from the original PoC...